File: FunctionTestRunner/FunctionTestRunner/Utils/ScenarioPropertyBag.cs
using Newtonsoft.Json;

namespace FunctionTestRunner.Utils;

public class ScenarioPropertyBag
{
    private readonly Dictionary<string, Tuple<bool, string>> _bag;

    public static ScenarioPropertyBag Create()
    {
        return new ScenarioPropertyBag();
    }

    private ScenarioPropertyBag()
    {
        _bag = new Dictionary<string, Tuple<bool, string>>();
    }

    public void Require(params string[] keys)
    {
        var missingKeys = new List<string>();
        foreach (var key in keys)
        {
            if (!_bag.ContainsKey(key))
            {
                missingKeys.Add(key);
            }
        }

        if (missingKeys.Count > 0)
        {
            throw new Exception($"The following data item(s) are not set in data bag: {string.Join(", ", missingKeys)}. Scenario is aborted.");
        }
    }

    public bool ContainsKeys(params string[] keys)
    {
        var missingKeys = new List<string>();
        foreach (var key in keys)
        {
            if (!_bag.ContainsKey(key))
            {
                missingKeys.Add(key);
            }
        }

        return missingKeys.Count == 0;
    }

    public bool ContainsKeys(params DataBagKey[] keys)
    {
        return ContainsKeys(keys.Select(x => x.ToString()).ToArray());
    }

    public T Get<T>(string key)
    {
        if (!_bag.ContainsKey(key) || !_bag[key].Item1)
        {
            return default(T);
        }

        return JsonConvert.DeserializeObject<T>(_bag[key].Item2);
    }

    public void Set(string key, object value)
    {
        var insert = new Tuple<bool, string>(value != null, value == null ? string.Empty : JsonConvert.SerializeObject(value));
        if (_bag.ContainsKey(key))
        {
            _bag[key] = insert;
        }
        else
        {
            _bag.Add(key, insert);
        }
    }

    public void RemoveKey(string key)
    {
        if (_bag.ContainsKey(key))
        {
            _bag.Remove(key);
        }
    }

    internal object Get<T>(object labelId)
    {
        throw new NotImplementedException();
    }
}

File: FunctionTestRunnerTS/FunctionTestRunner/src/utils/scenario-property-bag.ts
export class ScenarioPropertyBag {
    private bag: { [id: string] : [boolean, string]; };

    static create(): ScenarioPropertyBag {
        return new ScenarioPropertyBag();
    }

    constructor() {
        this.bag = {};
    }

    require(keys: string[]) {
        const missingKeys: string[] = [];

        keys.forEach(key => {
            if (!(key in this.bag)) {
                missingKeys.push(key);
            }
        });

        if (missingKeys.length > 0) {
            throw new Error(`The following data item(s) are not set in data bag: ${missingKeys.join(', ')}. Scenario is aborted.`);
        }
    }

    containsKeys(keys: string[]): boolean {
        var missingKeys: string[] = [];

        keys.forEach(key => {
            if (!(key in this.bag)) {
                missingKeys.push(key);
            }
        });

        return missingKeys.length == 0;
    }

    get<T>(key: string): T | undefined
    {
        if (!(key in this.bag) || !this.bag[key][0]) {
            return undefined;
        }

        return JSON.parse(this.bag[key][1]);
    }

    set(key: string, value: unknown) {
        const insert: [ boolean, string] = [ value !== null, value == null ? '' : JSON.stringify(value) ]

        this.bag[key] = insert;
    }

    removeKey(key: string) {
        if (key in this.bag) {
            delete this.bag[key]
        }
    }
}